### ISSUE TITLE
Change time step?

### ISSUE THREAD

**Issue by rgcoe** [PMEC-OSU/FOSTWIN#17]
From today's information session: Can we change the time step?

**Comment by johannes-spinneken** [comment]
There are two parts to this question:
- Can the solver type (fixed step vs. variable step) change?
- Can the time step change?

The first part is easy to address: the solver type needs to be fixed-step, as all common real-time systems only support fixed time steps. A variable time-step solver may lead to finer and finer time steps, for example changing from a time step of 5 ms to 0.5 ms when an end-stop is reached. Within a real-time system, computing each time step within a known, consistent, and deterministic time step is key. As a result, real-time systems such as Simulink Real-Time running on the Speedgoat target only support fixed time steps.

With this fixed time step in mind, a compromise must now be found between:
- a sufficiently small time step to ensure that the control system latency remains low
- a sufficiently large time step to ensure that the target hardware (Speedgoat) can solve the system within the bounds of the time step

For the FOSTWIN setup, we are using a [Speedgoat Baseline](https://www.speedgoat.com/products-services/real-time-target-machines/baseline-real-time-target-machine) system. This system can reach up to ~10kHz real-time behavior for **very** simple systems, but considerably less for more complex systems. For example, for the WecSim twin variant which relies on Simscape Multi-body and hence a 6 DoF solver, approx. 200 Hz (5ms) is the fasted the Speedgoat Baseline can support. For the SID model, which only requires computing a simple transfer function, 1000 Hz (1ms) can be reached.

In terms of the minimum time step required, considering the fastest dynamics of the model is key. For example, if a drive train resonance occurs at 50 Hz (20 ms), then 180 deg of this signal corresponds to 10 ms. To avoid introducing resonance (via sign inversion of the feedback signal by a 180 deg phase shift), the loop time needs to be << 10ms.  

For the present FOSTWIN example, a loop rate of 1000 Hz (1ms) is considered a good compromise between the various constraints. Indeed, 1000 Hz is a very common real-time loop rate for model testing at scale ~1:20 ... 1:50. For larger model tests (e.g. 1:5) or ocean-going devices, a real-time loop rate of 200 ... 500 Hz is common.

**To summarize the initial question:**
- _Can the solver type (fixed step vs. variable step) change?_ No: needs to be fixed step.
- _Can the time step change?_ For the purpose of the FOSTWIN control competition, 1ms is considered a good compromise between the various constraints, and this number should not be changed.
